### PR TITLE
[ws-manager-mk2] Support backup of workspaces

### PR DESCRIPTION
**Pull request by Furisto**
## Description
Workspaces that have been started by ws-manager-mk2 can now be stopped and successfully backed up.

## Related Issue(s)
n.a.

## How to test
- Open workspace in [preview environment](https://fo-mk2-dispose-ws.preview.gitpod-dev.com/workspaces)
- Stop workspace

## Release Notes

```release-note
None
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all-ci
- [ ] /werft no-test
      Run Leeway with `--dont-test`

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Review by WVerlaek — COMMENTED** [review]
Some small comments/questions - lgtm otherwise

File: components/ws-daemon/pkg/controller/controller.go
Comment (Furisto) [reply components/ws-daemon/pkg/controller/controller.go:182]: If the workspace CR has been modified in the meantime by someone else the update will be rejected (the CR generation will be lower than what is saved in etcd) . That's why we need to fetch the workspace again before we update (and after each failed attempt). It is a good point though, I will add a comment to make this clearer.

Comment (WVerlaek) [reply components/ws-daemon/pkg/controller/controller.go:182]: yeah - what i mean is that instead of storing the newer version in the locally scoped `workspace` var, we should load the new version in `ws`, such that when the outer function returns, anything that follows is also immediately using the newer version of the resource

Comment (Furisto) [reply components/ws-daemon/pkg/controller/controller.go:182]: D'oh! Of course. I will change it. :smile:

Comment (WVerlaek) [reply components/ws-daemon/pkg/controller/controller.go:182]: 👍

File: components/ws-daemon/pkg/controller/workspace_operations.go
Comment (WVerlaek) [inline components/ws-daemon/pkg/controller/workspace_operations.go:294]: still slightly prone to a race condition, what about setting a boolean if we reach the `case wso.backupWorkspaceLimiter <- struct{}{}:` case, and return based on that? Or otherwise, lets make `15` a const, so we don't forget to update the other if we change either

File: components/ws-daemon/pkg/controller/workspace_operations.go
Comment (WVerlaek) [inline components/ws-daemon/pkg/controller/workspace_operations.go:326]: thought I left a comment but got lost, but this should probably also be done inside the retry func if there was an error, otherwise only the last retry will get cleaned up

**Comment by Furisto** [comment]
@WVerlaek PTAL

File: components/ws-daemon/pkg/controller/workspace_operations.go
Comment (WVerlaek) [inline components/ws-daemon/pkg/controller/workspace_operations.go:333]: this cleanup should also still run on the last (successful) attempt, adding back the `defer` above `retryIfErr` should do it 😄

**Review by WVerlaek — APPROVED** [review]
LGTM, one final comment, but approving so you can merge when handled (or if done in a follow up PR).

/hold

**Pull request merged**
